File: packages/guardian-rn/android/src/main/kotlin/com/guardian/rn/ThreatBus.kt
package com.guardian.rn

import com.guardian.rn.generated.ThreatId
import com.guardian.rn.generated.Severity
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import java.util.concurrent.atomic.AtomicLong
import org.json.JSONObject

data class ThreatPayload(
    val threatId: String,
    val severity: String,
    val confidence: Double,
    val evidence: Map<String, String>,
    val ts: Long,
)

data class ThreatEnvelope(
    val seq: Long,
    val sessionId: String,
    val ts: Long,
    val hmac: String,
    val payload: ThreatPayload,
)

/**
 * Native threat event bus for Android.
 * Maintains a monotonic sequence number per ADR-0003.
 * Events are emitted as HMAC-signed envelopes onto a SharedFlow.
 * Per ADR-0004: replay = 32, suspend on buffer overflow.
 */
class ThreatBus(
    private val sessionId: String,
    private val sessionKey: ByteArray,
) {
    private val seq = AtomicLong(0L)

    private val _flow = MutableSharedFlow<ThreatEnvelope>(
        replay = 32,
        extraBufferCapacity = 64,
    )
    val flow: SharedFlow<ThreatEnvelope> = _flow.asSharedFlow()

    /**
     * Sign and emit a threat payload.
     * Increments the sequence number atomically.
     */
    suspend fun emit(payload: ThreatPayload) {
        val currentSeq = seq.incrementAndGet()
        val canonical = CanonicalJsonSerializer.canonicalize(
            mapOf(
                "threatId"   to payload.threatId,
                "severity"   to payload.severity,
                "confidence" to payload.confidence,
                "evidence"   to payload.evidence,
                "ts"         to payload.ts,
            )
        )
        val hmac = HmacSigner.sign(canonical, sessionKey)
        val envelope = ThreatEnvelope(
            seq       = currentSeq,
            sessionId = sessionId,
            ts        = System.currentTimeMillis(),
            hmac      = hmac,
            payload   = payload,
        )
        _flow.emit(envelope)
    }

    fun toJson(envelope: ThreatEnvelope): String {
        val evidenceJson = JSONObject(envelope.payload.evidence)
        val payloadJson = JSONObject().apply {
            put("threatId",   envelope.payload.threatId)
            put("severity",   envelope.payload.severity)
            put("confidence", envelope.payload.confidence)
            put("evidence",   evidenceJson)
            put("ts",         envelope.payload.ts)
        }
        return JSONObject().apply {
            put("seq",       envelope.seq)
            put("sessionId", envelope.sessionId)
            put("ts",        envelope.ts)
            put("hmac",      envelope.hmac)
            put("payload",   payloadJson)
        }.toString()
    }
}

File: packages/guardian-rn/ios/Sources/GuardianRN/ThreatBus.swift
import Foundation

struct ThreatPayload: Encodable {
    let threatId: String
    let severity: String
    let confidence: Double
    let evidence: [String: String]
    let ts: Int64
}

struct ThreatEnvelope: Encodable {
    let seq: Int64
    let sessionId: String
    let ts: Int64
    let hmac: String
    let payload: ThreatPayload
}

/// Native threat event bus for iOS.
/// Emits HMAC-signed envelopes via an AsyncStream (ADR-0004).
/// Sequence numbers are maintained with an os_unfair_lock-guarded counter.
final class ThreatBus {

    private let sessionId: String
    private let sessionKey: [UInt8]

    private var seq: Int64 = 0
    private var seqLock = os_unfair_lock_s()

    private var continuations: [String: AsyncStream<ThreatEnvelope>.Continuation] = [:]
    private var continuationLock = os_unfair_lock_s()

    init(sessionId: String, sessionKey: [UInt8]) {
        self.sessionId = sessionId
        self.sessionKey = sessionKey
    }

    /// Returns an AsyncStream of signed envelopes and a token to cancel it.
    func stream() -> (AsyncStream<ThreatEnvelope>, String) {
        let id = UUID().uuidString
        var cont: AsyncStream<ThreatEnvelope>.Continuation!
        let stream = AsyncStream(ThreatEnvelope.self, bufferingPolicy: .bufferingNewest(32)) { continuation in
            cont = continuation
        }
        os_unfair_lock_lock(&continuationLock)
        continuations[id] = cont
        os_unfair_lock_unlock(&continuationLock)
        return (stream, id)
    }

    func cancelStream(id: String) {
        os_unfair_lock_lock(&continuationLock)
        continuations[id]?.finish()
        continuations.removeValue(forKey: id)
        os_unfair_lock_unlock(&continuationLock)
    }

    /// Sign and emit a threat payload to all active streams.
    func emit(_ payload: ThreatPayload) throws {
        os_unfair_lock_lock(&seqLock)
        seq &+= 1
        let currentSeq = seq
        os_unfair_lock_unlock(&seqLock)

        let evidenceMap: [String: Any?] = payload.evidence.mapValues { $0 as Any? }
        let payloadMap: [String: Any?] = [
            "threatId":   payload.threatId,
            "severity":   payload.severity,
            "confidence": payload.confidence,
            "evidence":   evidenceMap,
            "ts":         payload.ts,
        ]
        let canonical = try CanonicalJSONEncoder.canonicalize(payloadMap)
        let hmacValue = HmacSigner.sign(canonicalPayload: canonical, keyBytes: sessionKey)

        let envelope = ThreatEnvelope(
            seq:       currentSeq,
            sessionId: sessionId,
            ts:        Int64(Date().timeIntervalSince1970 * 1000),
            hmac:      hmacValue,
            payload:   payload
        )

        os_unfair_lock_lock(&continuationLock)
        let snapshot = continuations
        os_unfair_lock_unlock(&continuationLock)

        for (_, cont) in snapshot {
            cont.yield(envelope)
        }
    }

    func toJSON(_ envelope: ThreatEnvelope) throws -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = []
        let data = try encoder.encode(envelope)
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}
